### PR TITLE
🤖 fix: checkpoint workflow agent reservations

### PR DESCRIPTION
**Pull request by ThomasK33**
## Summary

Checkpoints workflow-owned agent reservation before child task creation, so a workflow that enters `agent(...)` can no longer sit visibly as only a phase event while task reservation or launch work stalls.

## Background

Two workflow runs were observed stuck after their first phase with empty workflow steps and no child workspace. The sequential workflow agent path previously waited inside `TaskService.create(...)` before any durable workflow step/task state existed, leaving operators with no actionable boundary.

## Implementation

- Added a durable `agent-step` workflow event with `reserving`, `reserved`, and `failed` statuses.
- Routed sequential workflow `agent(...)` through `createAgentTasks([spec])` plus `waitForAgentTask(taskId, ...)` when the adapter supports reservation-first primitives.
- Recorded reservation events before calling task creation, and recorded failed reservation events if task reservation throws.
- Added schema, workflow details rendering, sidebar projection, Storybook fixture, and task_list/task_await progress coverage for reservation-only workflow progress.

## Validation

- `bun test src/node/services/workflows/WorkflowRunner.test.ts -t 'records an agent reservation event before reserving a single agent task|records a failed agent reservation event when reservation fails'`
- `bun test src/common/orpc/schemas/workflow.test.ts`
- `bun test src/browser/features/Tools/WorkflowRunToolCall.test.tsx`
- `bun test src/node/services/tools/task_list.test.ts -t 'reports agent reservation events as workflow progress before a child task exists' src/node/services/tools/task_await.test.ts -t 'surfaces reservation-only workflow progress while awaiting a run'`
- `git diff --check`
- `make typecheck`
- `MUX_ESLINT_CONCURRENCY=1 make static-check`
- Dogfooded the current-source workflow CLI path using the Coder AI Gateway bridge session token; the real workflow run completed and wrote the expected workflow events.

## Risks

Medium risk in workflow runner/event projection logic: this touches workflow agent lifecycle bookkeeping and UI rendering of workflow events. The change intentionally preserves existing task events and step records, and adds a separate event type for the no-task-id reservation boundary instead of weakening `task` event invariants.

## Pains

- Broad `WorkflowRunner.test.ts` filters can trigger Bun segfault/illegal-instruction failures locally, so targeted single-test filters were used for deterministic validation.
- The isolated dogfood MUX_ROOT needed the Coder AI Gateway session token paired as `OPENAI_API_KEY`; standard API keys against the bridge base URL produced 403s.

---

<details>
<summary>📋 Implementation Plan</summary>

# Implementation Plan: Durable workflow-owned agent reservation and launch

## Context and diagnosis

Two workflow runs (`1f65cd0812/wfr_cd1fc2b2daf8fa77` and `8fd91692cb/wfr_1bf8eccfa0b85b37`) were observed as “stuck” after their first workflow phase and before launching visible sub-agents.

Verified facts from the persisted session/workflow data:

- `1f65cd0812/wfr_cd1fc2b2daf8fa77`
  - Status: `running`.
  - Latest event: `phase("branch-recon")`.
  - `steps.jsonl`: empty.
  - No child workspace/config entry with `workflowTask.runId = "wfr_cd1fc2b2daf8fa77"`.
  - Lease was still being renewed, so the runner was live rather than orphaned.
- `8fd91692cb/wfr_1bf8eccfa0b85b37`
  - Status: `running`.
  - Latest event: `phase("direction-scan")`.
  - `steps.jsonl`: empty.
  - No child workspace/config entry with `workflowTask.runId = "wfr_1bf8eccfa0b85b37"`.
  - Lease was still being renewed.
- Both workflow source snapshots call a single sequential `agent(...)` immediately after the recorded phase:
  - `branch.js`: `agent(..., { id: "branch-recon", agentId: "explore", isolation: "none", schema: ... })`
  - `direction.js`: `agent(..., { id: "direction-scan", agentId: "explore", isolation: "none", schema: ... })`

Current single-agent launch path:

```text
WorkflowRunner.__workflowAgent
  -> WorkflowRunner.runOrResumeAgentStep
    -> WorkflowTaskServiceAdapter.runAgent
      -> TaskService.create
        -> returns taskId only after create/start work completes far enough
    -> WorkflowRunner.recordStepStarted(...taskId...)
```

The workflow has a black-hole window: it can be inside `TaskService.create(...)` while no workflow step, task event, or child workspace has been durably recorded. Lease renewal proves the runner is alive, but not that it is making progress.

## Goals

1. A workflow-owned `agent(...)` step must become visible and durable before any slow launch/startup work can block.
2. Workflow replay/resume must be keyed by durable `stepId + inputHash + taskId`, not by an in-memory promise.
3. A slow or hung child launch must produce actionable workflow state instead of a phase-only “running” card.
4. Single sequential `agent(...)`, `parallel(...)`, and `pipeline(...)` launches should use the same reservation-first semantics.
5. Task creation must avoid holding global task locks across slow fork/init/send-message work.
6. Interrupt/resume/restart behavior must remain deterministic and must not duplicate workflow-owned child tasks.

## Non-goals

- Do not redesign workflow authoring APIs (`agent`, `parallel`, `pipeline`) for users.
- Do not change workflow script replay keys except where needed to safely represent reservation state.
- Do not remove existing `TaskService.create(...)` behavior for non-workflow callers in the first slice; keep compatibility and route it through safer internals later.
- Do not solve unrelated stuck sub-agent/model/provider streaming issues.

## Recommended approach

Implement this as a staged **reservation-first workflow agent launch** refactor.

### Product-code LoC estimate

- **Recommended staged approach:** ~650–1,050 net product LoC.
  - Phase 1 visibility + single-agent reservation path: ~180–320 net product LoC.
  - Phase 2 shared reservation/launch API in `TaskService`: ~300–500 net product LoC.
  - Phase 3 reconciliation/diagnostics/UI polish: ~170–230 net product LoC.
- Tests are intentionally excluded from the product LoC estimate.

### Why staged?

Phase 1 should eliminate the observed black-hole bug quickly by routing single workflow agents through the already-existing `createAgentTasks([spec]) + waitForAgentTask(...)` machinery. Phase 2 then removes the underlying architectural trap by making task reservation first-class and moving slow work out of the task mutex. Phase 3 makes the state observable and repairable across crashes/restarts.

<details>
<summary>Alternatives considered</summary>

### Alternative A: Add logs only

- Product LoC: ~20–60.
- Rejected because it improves diagnosis but not recovery. A workflow could still be alive with no durable child-step state.

### Alternative B: Only add a workflow `creating` event before `TaskService.create(...)`

- Product LoC: ~80–160.
- Partially useful for UI visibility, but still leaves replay without a task id and still allows long awaits inside monolithic task creation.

### Alternative C: Full task-service refactor first

- Product LoC: ~700–1,100 in one slice.
- Architecturally clean but higher-risk. Prefer using the existing `createMany`/reservation shape first, then centralizing internals.

</details>

### Workflow-owned agent lifecycle

```text
1. parse/normalize agent spec
2. compute step inputHash
3. emit/record "reserving agent" state for visibility
4. reserve child task id durably
5. record workflow step with taskId
6. emit workflow task event: status queued|starting|started
7. launch/resume child task asynchronously
8. wait for `agent_report`/Plan proposal by taskId
9. record completed/failed/interrupted step result
```

### Durable state model

Prefer adding an explicit state that can represent “we are trying to reserve a task, but no task id exists yet.” There are two viable encodings; choose the least invasive after inspecting UI consumers.

#### Option 1 — Add a workflow event type: `agent-step`

Example:

```json
{
  "sequence": 3,
  "type": "agent-step",
  "at": "2026-07-01T15:33:55.500Z",
  "stepId": "branch-recon",
  "inputHash": "...",
  "status": "reserving",
  "title": "Recon branch diff"
}
```

Then later:

```json
{
  "sequence": 4,
  "type": "task",
  "at": "...",
  "stepId": "branch-recon",
  "taskId": "abc123",
  "status": "starting",
  "title": "Recon branch diff"
}
```

Pros:
- Does not overload existing `task` events that require `taskId`.
- Keeps “reservation without task id” visibly separate from “task exists.”

Cons:
- Requires UI/event rendering updates for a new event type.

#### Option 2 — Extend `WorkflowStepStatusSchema` with `reserving`

Example step record:

```json
{
  "stepId": "branch-recon",
  "inputHash": "...",
  "status": "reserving",
  "startedAt": "..."
}
```

Later transition to:

```json
{
  "stepId": "branch-recon",
  "inputHash": "...",
  "status": "started",
  "taskId": "abc123",
  "startedAt": "..."
}
```

Pros:
- Step-centric and useful for replay/reconciliation.

Cons:
- Requires careful migration semantics for `getStep(...)`, `recordStepStarted(...)`, and existing consumers that assume only `started|completed|failed|interrupted`.

**Default recommendation:** start with Option 1 for visibility and Phase 1 safety, then add step-status support only if reconciliation needs it. The core recovery boundary is the durable `taskId`, so avoid expanding step-state semantics unless necessary.

### Intent

Create tests that can fail on the current bug: a workflow enters `agent(...)` and remains visible only as a phase event with empty steps.

### Implementation notes

Add/extend tests near:

- `src/node/services/workflows/WorkflowRunner.test.ts`
- `src/node/services/workflows/WorkflowTaskServiceAdapter.test.ts` if present or worthwhile
- `src/node/services/taskService.test.ts` only for task reservation behavior that cannot be isolated at `WorkflowRunner`
- `src/common/orpc/schemas/workflow.test.ts` if schemas change

### Test cases

1. **Single workflow agent records visible state before child result**
   - Fake adapter delays `waitForAgentTask` indefinitely or until test-controlled release.
   - Run workflow: `phase("before"); agent("prompt", { id: "slow", schema })`.
   - Assert the run has an agent-step/task event or workflow step state before the child result resolves.

2. **Single workflow agent uses reservation path**
   - Fake adapter exposes `createAgentTasks` and `waitForAgentTask` spies.
   - Assert single `agent(...)` calls `createAgentTasks([spec])` and then waits by task id.
   - Assert it does not call the legacy blocking `runAgent` path when nonblocking primitives are available.

3. **Reservation failure is visible and terminal**
   - Fake `createAgentTasks` rejects or returns an error.
   - Assert workflow records an error event and failed status with the step id/title included.

4. **Replay with existing task id does not duplicate**
   - Seed a started step with `taskId` and matching `inputHash`.
   - Resume workflow.
   - Assert no new child is created; workflow calls `waitForAgentTask(existingTaskId, ...)`.

5. **Missing reserved child repair path**
   - Seed a started step with a task id, but fake `waitForAgentTask` throws `Task not found`.
   - Assert workflow records failed/interrupted terminal event or retries once according to existing `shouldRestartUnrecoverableStartedTask` behavior.
   - Ensure no unbounded duplicate creation loop.

6. **Contention visibility regression**
   - Simulate unrelated legacy task launch work holding or delaying the task creation path.
   - Assert a workflow entering `agent(...)` records `agent-step: reserving` immediately, before waiting on reservation.
   - After Phase 2, add/extend the test so an unrelated slow launch no longer blocks workflow task reservation itself.

7. **Reservation callback before config persistence**
   - Simulate `createMany`/reservation lifecycle returning a task id before the child config entry is durably written, then fail the config write.
   - Assert replay either repairs by retrying once or fails deterministically with a clear stale-reservation error; it must not loop or silently duplicate children.

### Quality gate

Run targeted tests and confirm at least one new test fails before production changes when feasible.

### Files/symbols likely involved

- `src/node/services/workflows/WorkflowRunner.ts`
  - `runOrResumeAgentStep(...)`
  - `runAndRecordAgentStepWithRetries(...)`
  - `recordTaskStartedEventIfMissing(...)`
  - `recordAgentResult(...)`
- `src/node/services/workflows/WorkflowTaskServiceAdapter.ts`
  - `createAgentTasks(...)`
  - `waitForAgentTask(...)`
  - `runAgent(...)`

### Steps

1. In `WorkflowRunner.runOrResumeAgentStep(...)`, prefer the nonblocking reservation path whenever `taskAdapter.createAgentTasks` and `taskAdapter.waitForAgentTask` are available.
2. Before calling `createAgentTasks`, emit a visible durable reservation event/log for the step.
   - This must happen **before** `createAgentTasks([resultSpec])`, because `createAgentTasks`/`createMany` can itself wait on task mutexes or validation work.
   - Keep this event distinct from a task event if no `taskId` exists yet.
   - Include `stepId`, `inputHash`, `title`, `agentId`, and `isolation` if schema permits those details.
3. Call `createAgentTasks([resultSpec], lifecycle)`.
   - In `onTaskCreated`/reservation lifecycle, immediately call `recordStepStarted(...)` and `recordTaskStartedEventIfMissing(...)` with the reserved task id.
   - Treat this as a **workflow-checkpointed reservation**, not proof that the child workspace has already materialized or started streaming.
   - Assert the returned task id matches the lifecycle task id if both are present.
4. Call `waitForAgentTask(taskId, resultSpec, waitOptions)`.
5. Keep existing retry behavior for schema validation failures.
6. Keep existing timeout-specific path intact initially, but verify whether it already uses `createAgentTasks`; if so, align the event/status naming.
7. Keep fallback to `runAgent(...)` only for adapters that do not support `createAgentTasks`/`waitForAgentTask`, and add a comment that production workflow adapters should not use this fallback.

### Acceptance criteria

- A single sequential workflow `agent(...)` creates a durable task id before waiting for child completion.
- A delayed child report no longer leaves the run with only a phase event; the run has an `agent-step` reservation event, a task event, or a step record.
- A blocked `waitForAgentTask` is visible as a workflow step with task id.
- Existing parallel/pipeline workflow tests still pass.

### Product LoC estimate

~120–220 net product LoC.

### Files/symbols likely involved

- `src/common/orpc/schemas/workflow.ts`
  - `WorkflowRunEventSchema`
  - `WorkflowStepStatusSchema` only if needed
- `src/common/types/workflow.ts`
- Workflow run card/event rendering under `src/browser/**workflow**` or related run UI components
- `src/node/services/workflows/WorkflowRunStore.ts`

### Steps

1. Add a representation for “agent step reservation started.” Preferred: a new workflow event type such as `agent-step` with status `reserving|reserved|failed`.
2. Update run/event rendering so users see something like:
   - `Reserving child task: branch-recon`
   - `Starting child task: branch-recon (abc123)`
3. Ensure event parsing remains backward compatible for existing workflow runs.
4. Avoid native `title` tooltips; use shared tooltip components if UI affordances need hover text.
5. If the UI changes the workflow run card layout, verify mobile/narrow width behavior.

### Acceptance criteria

- The UI distinguishes “workflow is between phase and task reservation” from “workflow is waiting on an existing child task.”
- Existing persisted runs parse successfully.
- New events appear in workflow details without breaking stream subscriptions.

### Product LoC estimate

~60–120 net product LoC.

### Files/symbols likely involved

- `src/node/services/taskService.ts`
  - `create(...)`
  - `createMany(...)`
  - `scheduleReservedTaskLaunch(...)`
  - `startReservedAgentTask(...)` / equivalent launch path
  - task queue/recovery helpers
- `src/node/services/workflows/WorkflowTaskServiceAdapter.ts`

### Design

Create a first-class internal API:

```ts
reserveAgentTasks(argsList, options): Promise<TaskCreateResult[]>
launchReservedAgentTask(taskId): Promise<void>
```

Or equivalently:

```ts
createReservedAgentTasks(argsList, options): Promise<TaskCreateResult[]>
```

with launch always handled by the scheduler.

### Steps

1. Extract the common validation/allocation/config-persistence portion from `createMany(...)` into a shared reservation helper.
2. Move potentially slow runtime/filesystem work outside the global task mutex where safe:
   - runtime construction that can touch external state
   - agent discovery/frontmatter resolution, or make it bounded and instrumented if it must remain coupled to reservation
   - fork/materialize workspace
   - init hooks
   - `WorkspaceService.sendMessage(...)`
   - model stream startup
3. Under the global task mutex, do only cheap config/counter mutation and short invariant checks:
   - re-read parent workspace/config state
   - verify trust/nesting/parallelism counters from current config
   - allocate `taskId` and workspace name
   - persist config workspace entry with `taskStatus`
   - emit or schedule metadata updates after persistence
4. Make `create(...)` call reservation + launch internally so external behavior stays compatible.
5. Make workflow adapter use reservation-first behavior directly, not the monolithic `create(...)` path.
6. Preserve the `onTaskReserved`/`onTaskCreated` lifecycle ordering guarantee needed by workflows:
   - workflows must checkpoint task id before slow launch starts
   - if config persistence fails after lifecycle callback, replay must see `Task not found` and repair/retry deterministically
7. Add assertions around task id presence and lifecycle consistency.

### Acceptance criteria

- Slow fork/init/send-message cannot block unrelated task reservation behind the global task mutex.
- `TaskService.create(...)` remains compatible for non-workflow callers.
- Workflow-owned child tasks always have a persisted parent workflow identity before launch.
- A crash after reservation but before launch is recoverable by queue/launch reconciliation.

### Product LoC estimate

~300–500 net product LoC.

### Files/symbols likely involved

- `src/node/services/workflows/WorkflowRunner.ts`
- `src/node/services/workflows/WorkflowRunStore.ts`
- `src/node/services/taskService.ts`
- `src/node/services/workspaceService.ts` only if workspace metadata/recovery hooks are needed
- workflow run UI components

### Reconciliation rules

| State | Child task exists? | Resume behavior |
|---|---:|---|
| no step/task event | no | reserve a new task |
| reservation event only, below timeout | no | continue waiting or retry reservation if runner restarted |
| reservation event only, stale | no | mark reservation failed and retry once or fail step with actionable error |
| step has `taskId` and status started | yes | wait/resume by task id |
| step has `taskId` and status started | no | record stale task, retry once with new task id or fail deterministically |
| step completed | yes/no | reuse completed result |
| step failed/interrupted | yes/no | respect explicit resume/retry policy |

### Steps

1. Add a bounded reservation diagnostic threshold, e.g. 30–60 seconds, for **no reservation progress** only.
   - Use it for warnings/actionable failure when no task id or reservation transition appears.
   - Do not fail legitimate queued/starting/running child tasks just because they have not completed or begun streaming.
2. If a workflow remains in reservation-only state past threshold, append a workflow log/error with step id and elapsed time.
3. On resume, detect stale reservation-only state and either retry reservation once or fail with a clear message.
4. Ensure `task_terminate` on a workflow run interrupts any reserved/started child tasks and cleans reservation state.
5. Ensure `task_list`/workflow status snapshots expose enough progress to distinguish active waiting from stale state.
6. Add lightweight debug logging at key boundaries, tagged with stable names (not noisy per-loop logs):
   - reservation requested
   - task id reserved
   - child launch scheduled
   - child wait started
   - child wait completed/failed

### Acceptance criteria

- Workflow runs cannot remain indefinitely `running` with only a phase event after entering `agent(...)`.
- Stale reservation states are visible and deterministic after restart.
- Interrupting a workflow with a reserved child does not leave an active orphan.
- Operators can identify the exact stuck boundary from workflow events without reading service logs.

### Product LoC estimate

~170–230 net product LoC.

### Steps

1. Remove any temporary debug instrumentation.
2. Add concise code comments at the reservation boundary explaining why workflows checkpoint task ids before launch.
3. Keep comments focused on the crash/replay invariant, not on the specific observed incident.
4. Avoid new free-floating docs unless explicitly requested.

### Product LoC estimate

~20–50 net product LoC.

### Targeted tests

Run focused tests first:

```bash
bun test src/node/services/workflows/WorkflowRunner.test.ts
bun test src/node/services/workflows/WorkflowTaskServiceAdapter.test.ts
bun test src/common/orpc/schemas/workflow.test.ts
bun test src/node/services/taskService.test.ts -t "workflow\|reserve\|queued\|starting"
```

Adjust exact test filenames/patterns after inspecting existing test layout.

### Static checks

```bash
make typecheck
MUX_ESLINT_CONCURRENCY=1 make static-check
```

Use `MUX_ESLINT_CONCURRENCY=1` if the workspace is memory constrained.

### Regression checks

- Existing workflow tests still pass.
- Existing task queue tests still pass.
- Existing schema compatibility tests still pass.
- No generated workflow source snapshots need migration.

### Setup

1. Start an isolated Mux dev environment using the project sandbox/dev-server workflow where available.
2. Use a temporary `MUX_ROOT` so dogfooding cannot mutate the user’s live `~/.mux` state.
3. Enable `dynamicWorkflows` and trust the test project.
4. Create or use a small test project with a workflow file that runs a slow workflow-owned agent.

Example dogfood workflow:

```js
export const meta = { description: "Slow workflow agent launch dogfood" };

export default function workflow({ phase, agent }) {
  phase("before-agent");
  const result = agent("Read the repo briefly, wait until you see this instruction, then return { ok: true } via structured report.", {
    id: "slow-agent",
    title: "Slow dogfood agent",
    agentId: "explore",
    isolation: "none",
    schema: {
      type: "object",
      required: ["ok"],
      additionalProperties: false,
      properties: { ok: { type: "boolean" } },
    },
  });
  return { reportMarkdown: "ok=" + result.ok, structuredOutput: result };
}
```

### Functional dogfood steps

1. Launch the workflow from the UI or via `workflow_run` in a sandbox workspace.
2. Immediately inspect the workflow card/details.
   - Expected: phase `before-agent` plus visible `reserving/starting/started` child step.
   - Expected: the run is no longer phase-only after entering `agent(...)`; it has an `agent-step` reservation event, task event, or step record.
3. Confirm a child workspace appears with matching `workflowTask.runId` and `workflowTask.stepId`.
4. Restart the sandbox Mux server while the child is active.
5. Resume/reopen the run.
   - Expected: same task id is reused; no duplicate child workspace appears.
6. Interrupt the workflow.
   - Expected: child task is interrupted or cleaned according to workflow-owned child rules.
7. Rerun the workflow to completion.
   - Expected: final result persists and replay returns cached/completed result without relaunching the child.

### Visual evidence

Capture and attach evidence for reviewers:

- Screenshot: workflow details immediately after reservation, showing the `reserving/starting` state.
- Screenshot: child workspace/sidebar entry with matching workflow step title.
- Screenshot: after restart/resume, showing the same task id/run continuing.
- Narrow/mobile viewport screenshot (~375px wide) if workflow UI layout changes.
- Short screen recording of launch → visible child step → restart/resume or interrupt.

If the environment is headless and video capture is unavailable, record the blocker and provide screenshots plus a terminal transcript produced with `script -q -e -c '<cmd>' /tmp/workflow-reservation.typescript`.

## Rollout strategy

1. Land Phase 1 and tests first if possible; it is the smallest high-value fix for the observed bug.
2. Land Phase 2 in the same PR only if the diff remains reviewable; otherwise split into a follow-up PR.
3. Keep old run parsing backward compatible.
4. Do not migrate historical workflow runs unless a schema change requires it; prefer optional fields/events.
5. After merge, monitor for:
   - workflows with `running` status and no post-phase step/event for > 60 seconds
   - duplicate workflow-owned child tasks for the same `runId + stepId + inputHash`
   - task queue stalls around reservation/launch

## Risks and mitigations

| Risk | Mitigation |
|---|---|
| Duplicate child tasks after crash/replay | Persist `taskId` before launch and make replay wait by existing task id. Add replay tests. |
| Existing UI assumes task events always have task ids | Use a separate `agent-step` event or keep `task` event unchanged. Add schema/UI tests. |
| Task reservation callback fires before config persistence | Existing replay should treat missing task as unrecoverable and retry/fail deterministically; add a targeted test for this edge. |
| Moving work out of task mutex changes queue ordering | Preserve reservation order and explicit `queued|starting` statuses under the mutex. Add queue tests. |
| Stale reservation failures become noisy | Use bounded diagnostics and one clear terminal failure; avoid repeating logs. |
| Workflow timeout semantics regress | Keep timeout path initially intact, then align after tests prove behavior. |

## Final acceptance criteria

- The original symptom is impossible: after a workflow enters `agent(...)`, it cannot remain indefinitely with only a phase event and empty steps.
- Single sequential `agent(...)` launches are durably visible before waiting for child output.
- Parallel/pipeline behavior remains compatible.
- Replay/resume never duplicates a child task for the same `stepId + inputHash` unless the previous reservation is proven unrecoverable and the retry is recorded.
- Interrupting a workflow cleans up or interrupts workflow-owned children.
- UI/API state identifies whether the workflow is reserving a task, waiting on a child, completed, failed, or interrupted.
- Targeted tests, `make typecheck`, and `make static-check` pass.
- Dogfooding evidence includes screenshots and a short recording or documented headless fallback.

## Advisor review status

Advisor review: Fully approved. No blocking changes remain. The final review confirmed the plan addresses pre-reservation observability, replay semantics, `createMany` lifecycle ambiguity, mutex boundaries, contention coverage, and avoiding false failures for normal queued/running children.

</details>

---

_Generated with [`mux`](https://github.com/coder/mux) • Model: `openai:gpt-5.5` • Thinking: `xhigh` • Cost: `1939195{MUX_COSTS_USD:-unknown}`_

### CODE REVIEW

**Comment by ThomasK33** [comment]
@codex review

File: src/browser/features/RightSidebar/Workflows/projectWorkflowRun.ts [hunk 1]
```diff
@@ -203,6 +203,7 @@ function buildPhaseView(
 // assignment keys off the first of ANY of these per stepId so every step lands in its phase.
 function stepBearingEventStepId(event: WorkflowRunEvent): string | null {
   switch (event.type) {
+    case "agent-step":
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/browser/features/RightSidebar/Workflows/projectWorkflowRun.ts:206]: Addressed in 3cb9d9d3b. The workflow run projector now synthesizes a running/failed timeline row from unmatched agent-step reservation events keyed by stepId+inputHash, so reservation-only workflow steps appear in the sidebar/header before a taskId or step record exists. Added projectWorkflowRun coverage for the reservation-only case.

**Comment by ThomasK33** [comment]
@codex review

Please take another look. I addressed the reservation-only sidebar projection finding and pushed the fix.
